specs/ocata/approved/hierarchy-quota-driver-api.rst

```diff
@@ -0,0 +1,294 @@
+   |          +--------+            |
+   |              /\                |
+   |             /  \               |
+   |            /    \              |
+   |           /      \             |
+   |          /        \            |
+   |    +---------+  +---------+    |
+   |    | Prj_0_a |  | Prj_0_b |    |
+   |    +---------+  +---------+    |
+   |    |quota=10 |  |quota=10 |    |
+   |    +---------+  +---------+    |
+   |          /\                    |
+   |         /  \                   |
+   |        /    \                  |
+   |       /      \                 |
+   |      /        \                |
+   | +---------+  +---------+       |
+   | | Prj_1_a |  | Prj_1_b |       |
+   | +---------+  +---------+       |
+   | |quota=7  |  |quota=10 |       |
+   | +---------+  +---------+       |
+   |                                |
+   +--------------------------------+
+
+   2.1
+
+   a. Try to allocate 8 items of resources for project "Prj_1_a" - get an
+      error due to the project limits.
+   b. Try to allocate 7 items of resources for project "Prj_1_a" - success.
+      Prj_1_a usage=7. Prj_0_a usage=7.
+   c. Try to allocate one more item for "Prj_1_a" - get an error due to the
+      project limits.
+   d. Try to allocate 3 items for "Prj_1_b" - success. Prj_1_b usage=3.
+      Prj_0_a usage=10.
+   e. Try to allocate one more item for project "Prj_1_b" - get an error due
+      to the parent project limits.
+
+   2.2
+
+   a. Try to allocate 5 items for project "Prj_0_a" - success.
+      Prj_0_a usage=5.
+   b. Try to allocate 5 items for project "Prj_1_a" - success.
+      Prj_1_a usage=5. Prj_0_a usage=10.
+   c. Try to allocate one more item for project "Prj_1_a" - get an error due
+      to the parent project limits.
+
+Project Priority
+----------------
+
+None
+
+
+Proposed change
+===============
+
+1. The main idea is to implement nested quotas without overbooking and
+   with overbooking in one Quota Driver.
+2. The default quota (hard limit) for any newly created project is set to 0.
+3. Each time when trying to allocate resources for project should be done check
+   of usages/limits.
+4. The limits enforcement should be done such way that not only project usage
+   can't exceed project limit but the sum of project usage and subprojects usages
+   can't exceed project limit.
+   Sum of subprojects limits shouls be <= project limit.
+5. In the overbooking mode sum of subprojects limits can be not equal to parent
+   project limit. Sum of subprojects limits can be greater then project limit. However,
+   sum of usages in subprojects can not ever exceed the quota limits for project.
+
+Alternatives
+------------
+
+
+Data model impact
+-----------------
+
+
+REST API impact
+---------------
+
+None
+
+
+Security impact
+---------------
+
+None
+
+
+Notifications impact
+--------------------
+
+None
+
+
+Other end user impact
+---------------------
+
+None
+
+
+Performance Impact
+------------------
+
+None
```

Comment:
Unknown and should be carefully measured because of changed approach to storing/measuring project limits and usages.